src/llm/io_processing/qwen3coder/qwen3coder_tool_parser.cpp

```diff
@@ -127,77 +114,115 @@ static const ParametersTypeMap_t parseToolSchema(const std::string& functionName
+        v.Accept(writer);
+        s = buffer.GetString();
+    }
+    v.SetString(s.c_str(), static_cast<rapidjson::SizeType>(s.size()), alloc);
+}
+
+void Qwen3CoderToolParserImpl::addParameterToCurrentFunctionDoc(std::string& parameterValueAsString) {
+    if (this->removeNewlineAroundParameters)
+        trimNewline(parameterValueAsString);
+    // now we have parameter value in string format. We need to use toolsSchemas to determine if it is string, number, bool, array or object
+    auto paramIt = this->toolsParametersTypeMap.find(this->currentFunction.name);
+    auto& currentFunctionArgsDoc = this->currentFunction.argumentsAsDocument;
+    auto& allocator = currentFunctionArgsDoc.GetAllocator();
+    auto& key = this->currentParameterName;
+    rapidjson::Value keyVal(key.c_str(), allocator);
+    rapidjson::Document temp;
+    if (paramIt != this->toolsParametersTypeMap.end()) {
+        auto paramJt = paramIt->second.find(currentParameterName);
+        if (paramJt != paramIt->second.end() && (paramJt->second == ParameterType::BOOLEAN)) {
+            std::transform(parameterValueAsString.begin(), parameterValueAsString.end(), parameterValueAsString.begin(), ::tolower);
+        }
+    }
+    temp.Parse(parameterValueAsString.c_str());
+    if (temp.HasParseError()) {
+        rapidjson::ParseErrorCode errorCode = temp.GetParseError();
+        size_t errorOffset = temp.GetErrorOffset();
+        SPDLOG_DEBUG("Error parsing parameter: {} value: {}; error at offset: {}; code: {};", this->currentParameterName, parameterValueAsString, errorOffset, rapidjson::GetParseError_En(errorCode));
+        // since we are unable to parse with rapidjson generated parameter will be inserted as is as string
+        rapidjson::Value v;
+        v.SetString(parameterValueAsString.c_str(), static_cast<rapidjson::SizeType>(parameterValueAsString.size()), allocator);
+        if (!currentFunctionArgsDoc.HasMember(keyVal)) {
+            currentFunctionArgsDoc.AddMember(keyVal, v, allocator);
+        } else {
+            SPDLOG_DEBUG("Parameter: {} already exists in document, overwriting with string value due to parse error", key);
+        }
+    } else {
+        rapidjson::Value valueCopy;
+        valueCopy.CopyFrom(temp, allocator);
+        if (paramIt != this->toolsParametersTypeMap.end()) {
+            auto paramJt = paramIt->second.find(currentParameterName);
+            if (paramJt != paramIt->second.end() && (paramJt->second == ParameterType::STRING)) {
+                enforceStringValue(valueCopy, allocator);
+            }
+        }
+        // Add or overwrite key in the main document
+        if (!currentFunctionArgsDoc.HasMember(keyVal)) {
+            SPDLOG_TRACE("Will add key:{} val:{} type:{}", key, parameterValueAsString, jsonTypeOf(valueCopy));
+            currentFunctionArgsDoc.AddMember(keyVal, valueCopy, allocator);
+        } else {
+            SPDLOG_DEBUG("Parameter: {} already exists in document.", key);
```

Comment:
Does it happen often? If yes then maybe we could check for already existing key earlier to avoid doing operations above if `valueCopy` is not required?